### PR TITLE
lib/mongo: introduce mongo queries

### PR DESCRIPTION
**Pull request by Morriar**
A small abstraction of JsonObject dedicated to repo queries.

Just because I'm fed up with writing my JsonObjects by hand...

Signed-off-by: Alexandre Terrasa <alexandre@moz-code.org>

### CODE REVIEW

File: lib/mongodb/queries.nit [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+
+	# Match documents where `field` is in `values`
+	#
+	# ~~~json
+	# { field: { $in: [<value1>, <value2>, ... <valueN> ] } }
+	# ~~~
+	#
+	# `$in` selects the documents where the value of a field equals any value
+	# in the specified array.
+	fun is_in(field: String, values: Array[nullable Jsonable]): MongoMatch do
```
Comment (jcbrinfo) [inline lib/mongodb/queries.nit]: Why don’t you use `op(…)` here?

File: lib/mongodb/queries.nit [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+	# Match documents where `field` is not in `values`
+	#
+	# ~~~json
+	# { field: { $nin: [<value1>, <value2>, ... <valueN> ] } }
+	# ~~~
+	#
+	# `$nin` selects the documents where:
+	# * the field value is not in the specified array or
+	# * the field does not exist.
+	fun is_nin(field: String, values: Array[nullable Jsonable]): MongoMatch do
```
Comment (jcbrinfo) [inline lib/mongodb/queries.nit]: Why don’t you use `op(…)` here?

File: lib/mongodb/queries.nit [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+#	gt("age", 18).
+#	exists("email", true).
+#	is_in("status", [1, 2, 3, 4])
+# ~~~
+class MongoMatch
+	super JsonObject
+
+	private fun op(name: String, field: String, value: nullable Jsonable): MongoMatch do
+		var q = new JsonObject
+		q["${name}"] = value
```
Comment (jcbrinfo) [inline lib/mongodb/queries.nit]: Sorry if this question sound stupid, I am not familiar with MongoDB. But, how a key string that starts with `$` is distinguished from a query? I mean, suppose we have the following document:

```
{
    "$ne": 42
}
```

How do you look for documents where `"$ne"` maps to `42`? In order words, how do you escape the meaning of `"$ne"` for MongoDB?

File: lib/mongodb/queries.nit [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+		self[field] = q
+		return self
+	end
+
+	# Match documents where `field` equals `value`
+	#
+	# ~~~json
+	# {field: {$eq: value} }
+	# ~~~
+	fun eq(field: String, value: nullable Jsonable): MongoMatch do
```
Comment (jcbrinfo) [inline lib/mongodb/queries.nit]: May I propose to add `match` as an alias to suggest that another `MongoMatch` may be used as the `value`.

File: lib/mongodb/queries.nit [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+# To this:
+#
+# ~~~nitish
+# collection.find((new MongoMatch).eq("login", "Morriar").exists("email", true))
+# ~~~
+#
+# The framework provides three classes:
+# * `MongoMatch` the base query that can be used with most Mongo services
+# * `MongoPipeline` the array of queries that is expected by `MongoCollection::aggregate`
+# * `MongoGroup` the group query for a `MongoPipeline`
```
Comment (privat) [inline lib/mongodb/queries.nit]: Not a lot of information in this 3 items.

File: lib/mongodb/queries.nit [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+	super JsonObject
+
+	private fun op(name: String, field: String, value: nullable Jsonable): MongoMatch do
+		var q = new JsonObject
+		q["${name}"] = value
+		self[field] = q
+		return self
+	end
+
+	# Match documents where `field` equals `value`
```
Comment (privat) [inline lib/mongodb/queries.nit]: Can `value` be a mach? If so then *equals* is not the right verb.

File: lib/mongodb/queries.nit [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+
+	# Match documents where `field` exists or not
+	#
+	# ~~~json
+	# {field: {$exists: value} }
+	# ~~~
+	#
+	# When `exists` is true, `$exists` matches the documents that contain the
+	# field, including documents where the field value is null.
+	# If <boolean> is false, the query returns only the documents that do not
```
Comment (privat) [inline lib/mongodb/queries.nit]: what is `<boolean>`?

File: lib/mongodb/queries.nit [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+		return self
+	end
+
+	# Match documents where `field` is in `values`
+	#
+	# ~~~json
+	# { field: { $in: [<value1>, <value2>, ... <valueN> ] } }
+	# ~~~
+	#
+	# `$in` selects the documents where the value of a field equals any value
```
Comment (privat) [inline lib/mongodb/queries.nit]: can `value` be a match? (cf. the comment above for `eq`)

File: lib/mongodb/queries.nit [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+		return self
+	end
+
+	# Match documents where `field` is greater or equal to `value`
+	#
+	# ~~~json
+	# {field: {$gte: value} }
+	# ~~~
+	fun gte(field: String, value: nullable Jsonable): MongoMatch do
+		op("gte", field, value)
```
Comment (privat) [inline lib/mongodb/queries.nit]: What kind of punk use `gte` for `<=`?

File: lib/mongodb/queries.nit [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+	end
+
+	# Push to an array.
+	#
+	# ~~~json
+	# { $push: <expression> }
+	# ~~~
+	#
+	# Returns an array of all values that result from applying an expression to
+	# each document in a group of documents that share the same group by key.
```
Comment (privat) [inline lib/mongodb/queries.nit]: errr... Do you have an example?

File: lib/mongodb/queries.nit [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+	#
+	# ~~~json
+	# {field: {$exists: value} }
+	# ~~~
+	#
+	# When `exists` is true, `$exists` matches the documents that contain the
+	# field, including documents where the field value is null.
+	# If <boolean> is false, the query returns only the documents that do not
+	# contain the field.
+	fun exists(field: String, value: nullable Jsonable): MongoMatch do
```
Comment (privat) [inline lib/mongodb/queries.nit]: I have an hard time with this one. Must `value` be either `true` or `false`? If yes, why not type with `Bool`.
And why not drop the `value` completely and add another method `not_exists`?

**Comment by privat** [comment]
some nitpicks but +1 anyway

File: lib/mongodb/queries.nit (hunk 3 above)
Comment (Morriar) [inline lib/mongodb/queries.nit]: With mongo:

~~~
> db.foo.save({"$ne": "lol"})
2016-08-14T12:34:45.552-0400 E QUERY    Error: field names cannot start with $ [$ne]
~~~

File: lib/mongodb/queries.nit (hunk 4 above)
Comment (Morriar) [inline lib/mongodb/queries.nit]: I will not add all the possible Jsonable variation of this method.

File: lib/mongodb/queries.nit (hunk 6 above)
Comment (Morriar) [inline lib/mongodb/queries.nit]: All the feature here are mapped from Mongo, I will not change the semantics.

File: lib/mongodb/queries.nit (hunk 9 above)
Comment (Morriar) [inline lib/mongodb/queries.nit]: Semantics of Monfo

**Comment by Morriar** [comment]
Reroll:
* aded link to the documentation of each feature from the MongoDB official documentation
* used `op` where possible
* use `Bool` argument for `exists`

File: lib/mongodb/queries.nit (hunk 3 above)
Comment (jcbrinfo) [inline lib/mongodb/queries.nit]: Seems legit.

File: lib/mongodb/queries.nit (hunk 9 above)
Comment (jcbrinfo) [inline lib/mongodb/queries.nit]: So the MongoDB team are kind of punks. :smile:

**Pull request merged**